controller/client/client.go

```diff
@@ -31,6 +32,89 @@ type Client struct {
 	*httpclient.Client
 }
 
+type JobWatcher struct {
+	events chan *ct.JobEvent
+	stream stream.Stream
+}
+
+func newJobWatcher(events chan *ct.JobEvent, stream stream.Stream) *JobWatcher {
+	w := &JobWatcher{
+		events: events,
+		stream: stream,
+	}
+	return w
+}
+
+func jobEventsEqual(expected, actual ct.JobEvents) bool {
+	for typ, events := range expected {
+		diff, ok := actual[typ]
+		if !ok {
+			if len(events) == 0 {
+				continue
+			}
+			return false
+		}
+		for state, count := range events {
+			actualCount, ok := diff[state]
+			if !ok || actualCount != count {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+func (w *JobWatcher) WaitFor(expected ct.JobEvents, timeout time.Duration, callback func(*ct.JobEvent) error) error {
+	actual := make(ct.JobEvents)
+	for {
+		select {
+		case e, ok := <-w.events:
+			if !ok {
+				if err := w.stream.Err(); err != nil {
+					return err
+				}
+				return fmt.Errorf("Event stream unexpectedly ended")
+			}
+			if _, ok := actual[e.Type]; !ok {
+				actual[e.Type] = make(map[string]int)
+			}
+			switch e.State {
+			case "crashed":
+				actual[e.Type]["down"] += 1
+			case "starting", "up", "down":
+				fallthrough
+			default:
+				actual[e.Type][e.State] += 1
+			}
+			if callback != nil {
+				err := callback(e)
+				if err != nil {
+					return err
+				}
+			}
+			if jobEventsEqual(expected, actual) {
+				return nil
+			}
+		case <-time.After(timeout):
+			return fmt.Errorf("Timed out waiting for job events. Waited %d seconds.\nexpected: %s\nactual: %s", timeout, toJson(expected), toJson(actual))
```

Comment:
I think the default printing of the `map` type (i.e. `%v`) rather than converting to JSON should suffice in the error output, e.g.:

```
map[web:map[up:1] worker:map[down:3]]
```